os/audio/audio.c

```diff
@@ -693,6 +693,42 @@ static inline void audio_dequeuebuffer(FAR struct audio_upperhalf_s *upper, FAR
 	}
 }
 
+/****************************************************************************
+ * Name: audio_handleerror
+ *
+ * Description:
+ *   Send an AUDIO_MSG_IOERR message to the client to indicate that the
+ *   an error has occured.  The lower-half driver initiates this
+ *   call via its callback pointer to our upper-half driver.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_AUDIO_MULTI_SESSION
+static inline void audio_handleerror(FAR struct audio_upperhalf_s *upper, FAR struct ap_buffer_s *apb, uint16_t status, FAR void *session)
+#else
+static inline void audio_handleerror(FAR struct audio_upperhalf_s *upper, FAR struct ap_buffer_s *apb, uint16_t status)
+#endif
+{
+	struct audio_msg_s msg;
+
+	audvdbg("Entry\n");
+
+	/* Send a error message to the user if a message queue is registered */
+
+	upper->started = false;
+	if (upper->usermq != NULL) {
+		/* We are always sending XRUN error msg since it is the
+		only type of error that we handle now. If new error scenario
+		is required, then we need to handle here */
+		msg.msgId = AUDIO_MSG_XRUN;
+		msg.u.pPtr = NULL;
+#ifdef CONFIG_AUDIO_MULTI_SESSION
+		msg.session = session;
+#endif
+		mq_send(upper->usermq, (FAR const char *)&msg, sizeof(msg), MQ_PRIO_MAX);
```

Comment:
since error notification happens at irq level. so this mq_send happens at interrupt context. would be great to know the following?
a. Did you test this funcitonality?
b. Did mq_send work properly from interrupt_context, any sideeffects of this?
b. Any rough esmitate on how much time it takes for the error notification from isr to reach the app queue?

Reply:
a. No, I have not tested. I need audio module to test.
b. Not sure if it will work. But, in my opinion, we must not be doing this. I suggest that we do it on a worker thread. This must be done either in I2S driver or codec driver. audio.c is just a wrapper file and must not create new worker.
c. Not sure. But I am using MQ_PRIO_MAX so that the error msg will take priority over all others in the audio msg queue

Reply:
if you think, "But, in my opinion, we must not be doing this." then please this as a NOTE in main review comment. But, it is still worth to get metrics/statistics even if we forbid this. 

Regarding your suggestion; 
Can you elaborate further to know if that is feasibile ?

Reply:
I think you can call the err_callback function from i2s_rx_worker or i2s_tx_worker. Since, we know that application only will respond to the xrun error, there is no point in handling it in interrupt context.

Reply:
In i2s file.. create new function like err_cb_worker().
In this function, just call priv->err_cb().
In i2s_irq_handler, schedule err_cb_worker() instead of calling priv->err_cb().

How about this?

Reply:
@btheosam What is your opinion?

Reply:
Rough estimate gives around 25ms delay for the error propogation from irq to user app (assuming mq_send is NOT interrupted & use app gets its time slice soon). With 8192 periods, this should be okay.